crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+                    // will do, no need for randomness here.
+                    assert!(!this_module.is_empty());
+                    let new_id = this_module.pop().expect("List should never be empty");
+                    if this_module.is_empty() {
+                        per_module.remove(&id.unwrap());
+                    }
+                    // Make sure to remove from the global
+                    // freelist. We already removed from the
+                    // per-module list above.
+                    Self::remove_free_list_item(slot_state, free_list, new_id);
+                    slot_state[new_id] = SlotState::Taken;
+                    new_id
+                } else {
+                    // Pick a random free slot ID. Note that we do
+                    // this, rather than pick a victim module first,
+                    // to maintain an unbiased stealing distribution:
+                    // we want the likelihood of our taking a slot
+                    // from some other module's freelist to be
+                    // proportional to that module's freelist
+                    // length. Or in other words, every *slot* should
+                    // be equally likely to be stolen. The
+                    // alternative, where we pick the victim module
+                    // freelist first, means that either a module with
+                    // an affinity freelist of one slot has the same
+                    // chances of losing that slot as one with a
+                    // hundred slots; or else we need a weighted
+                    // random choice among modules, which is just as
+                    // complex as this process.
+                    //
+                    // We don't bother picking an empty slot (no
+                    // established affinity) before a random slot,
+                    // because this is more complex, and in the steady
+                    // state, all slots will see at least one
+                    // instantiation very quickly, so there will never
+                    // (past an initial phase) be a slot with no
+                    // affinity.
+                    let free_list_index = rand::thread_rng().gen_range(0..free_list.len());
+                    let new_id = free_list[free_list_index];
+                    // Remove from both the global freelist and
+                    // per-module freelist, if any.
+                    Self::remove_free_list_item(slot_state, free_list, new_id);
+                    if let &SlotState::Affinity { module, .. } = &slot_state[new_id] {
+                        Self::remove_module_free_list_item(slot_state, per_module, module, new_id);
+                    }
+                    slot_state[new_id] = SlotState::Taken;
+
+                    new_id
+                }
+            }
+        }
+    }
+
+    pub(crate) fn free(&mut self, index: usize, id: Option<CompiledModuleId>) {
```

Comment:
I think the `id` is unnecessary here, and we can recover this from the `slot_state[index]`, if it exists? If so, removing `id` would help move us towards a single source of truth with fewer things that can get misaligned and out of sync, and would allow us to clean up slots that have a module affinity even when callers forgot to pass a module id here.

Reply:
Yes, that's a good point; made it a field of the `Taken` state now, to remember it for the free.